extensions/logging/log4j2_extensions/src/main/java/io/opentelemetry/contrib/logging/log4j2/TraceContextDataProvider.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.contrib.logging.log4j2;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Tracer;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.core.util.ContextDataProvider;
+
+/**
+ * This ContextDataProvider is loaded via the ServiceProvider facility. {@link #supplyContextData()}
+ * is called when a log entry is created.
+ */
+public class TraceContextDataProvider implements ContextDataProvider {
```

Comment:
Please investigate if possible to not call `toLowerBase16` on the critical path. Save the current SpanContext in the map and format the strings when exporting:
1. Remove unnecessary formatting and allocations from the critical path
2. Remove unnecessary formatting and allocations if log is dropped in the export pipeline.

Reply:
Unfortunately one of the main use cases here is use by the PatternLayout class, which won't know to call `toLowerbase16` when used, so this needs to be a string. On the positive side, this code is not called if the logging call is below the current threshold- I just confirmed that if a message is logged at a `trace` level, the message is discarded before the `TraceContextDataProvider` is invoked.